src/pins/gen.rs

```diff
@@ -103,7 +123,8 @@ macro_rules! pins {
         /// [`GPIO`] for more information.
         ///
         /// [`GPIO`]: ../gpio/struct.GPIO.html
-        pub struct Token<T, State>(T, PhantomData<State>);
+        // TODO impl getter instead of making T pub
+        pub struct Token<Trait: ?Sized, State>(pub Trait, PhantomData<State>);
```

Comment:
Why the `?Sized` bound? Why wouldn't we always know the size of `pins::Trait` implementations? This is also causing the CI failure.

I see no reason not to leave this as it was (besides `pub` or accessor method; of course, I might be missing something).

Reply:
The honest answer is: because rustc suggested it and the resulting error was more concise – but the "before" error afaik says the same:

```
error[E0277]: the size for values of type `(dyn pins::traits::Trait + 'static)` cannot be known at compilation time
   --> src/gpio.rs:386:16
    |
386 |         token: Token<dyn pins::Trait, init_state::Enabled>,
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
   ::: src/pins/gen.rs:131:26
    |
131 |         pub struct Token<Trait, State>(pub Trait, PhantomData<State>);
    |                          ----- required by this bound in `pins::gen::Token`
    |
    = help: the trait `Sized` is not implemented for `(dyn pins::traits::Trait + 'static)`
help: you could relax the implicit `Sized` bound on `Trait` if it were used through indirection like `&Trait` or `Box<Trait>`
   --> src/pins/gen.rs:131:26
    |
131 |           pub struct Token<Trait, State>(pub Trait, PhantomData<State>);
    |                            ^^^^^             ----- ...if indirection was used here: `Box<Trait>`
    |                            |
    |                            this could be changed to `Trait: ?Sized`...
...
169 | / pins!(
170 | |     pio0_0 , PIO0_0 , 0, 0x00, state::Unused;
171 | |     pio0_1 , PIO0_1 , 0, 0x01, state::Unused;
172 | |     pio0_2 , PIO0_2 , 0, 0x02, state::Swm<((),), ()>;
...   |
223 | |     pio1_21, PIO1_21, 1, 0x15, state::Unused;
224 | | );
    | |__- in this macro invocation
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

From my understanding, we can't know the size of all possible `pins::Trait` implementations as soon as the Trait contains methods instead of only consts. Our concrete implementation in `pins!` is known, but that doesn't make it hold true for the rest of the universe? Especially not if we re-add a `struct DynamicPin` that returns dynamic values for `pins::Trait`?

Reply:
We don't need to know the size of all possible `pins::Trait` implementations! So far, the API handles this using type parameters:

``` rust
fn use_pin<Pin>(pin: &mut Pin) {}
```
`Pin` is just a placeholder. Code for this function will only be generated, if you actually call it, and then the compiler will know the concrete type (and therefore its size). Yes, the calling function could itself have a generic `Pin`, but then someone is going to call that function (or some function up the chain) with a concrete type, or no code is ever going to be generated.

That we don't know the size of all possible `pins::Trait` implementations only becomes a problem once you make it one. Which you did with this bit (quoting your error message): `Token<dyn pins::Trait, init_state::Enabled>`. `dyn Trait` means "any implementation of trait", which the compiler can't know the size for. Since type parameters carry an implicit `Sized` bound, the compiler rightly complains that you can't use that here.

But the compiler's suggestion is misleading, because (as far as I can tell), there never was a good reason to use `dyn Trait` in the first place. Also, please note the following bit from the error message (emphasis mine):
> you could relax the implicit `Sized` bound on `Trait` **if it were used through indirection like `&Trait` or `Box<Trait>`**

Telling the compiler "this type doesn't need to have a known size" doesn't help, if the type is used for a struct field. There's no way to generate machine code from that. So it rightly tells you, that you also need indirection here, via `&Trait` or `Box<Trait>`. A pointer has a defined size, so we can put it in a struct. That the pointer points to something with an unknown size doesn't matter.

But in our case, I think this indirection is unnecessary. Using type parameters in combination with trait bounds everywhere, as before, should work fine.